prov/sockets/src/sock_ep.c

```diff
@@ -1795,12 +1795,19 @@ struct sock_conn *sock_ep_lookup_conn(struct sock_ep_attr *attr, fi_addr_t index
 {
 	int i;
 	uint16_t idx;
+	char buf[8];
 	struct sock_conn *conn;
 
 	idx = (attr->ep_type == FI_EP_MSG) ? index : index & attr->av->mask;
 
 	conn = ofi_idm_lookup(&attr->av_idm, idx);
 	if (conn && conn != SOCK_CM_CONN_IN_PROGRESS) {
+		if (conn->connected == 0)
+			return NULL;
+		if (sock_comm_peek(conn, buf, 8) == 0 && conn->connected == 0) {
```

Comment:
Have you looked at the performance impact of adding this to every lookup?

Reply:
Yes, i used fi_pingpong to test the ofi level msg rate over sockets provider over OPA's tcp/ip, it basically reports same number with or without the patch. DAOS level RPC throughput also be the same. thx

Reply:
I guess I don't fully understand the problem.  It seems that this just narrows a race somewhere.  What if the peer disconnects immediately after peek succeeds?  That needs to be later somewhere.  Either that code is missing, or there's some reason why handling it further down doesn't work.  Can you describe what's happening in more detail?

Reply:
Thanks for looking the issue, it is to handle the case that peer restarts executing.
For example two nodes A and B, A send RPC request to B, and B will send reply to A. To make B be able to send reply msg to A, B needs to insert A's address to AV to get the usable fi_addr. But B cannot insert it every time as it will consume an AV slot each time, so it only insert it at the first time and reuse it later.
Then A restart executing with same IP+port, and send new RPC request to B, B finds A's address already inserted to AV before and reuses the previous fi_addr. And in libfabric it also reuse the previous connection information, but that connection is broken and invalid. It cause the communication failure and cannot recover from that situation. The failure also happens for bulk handling case.

So this patch do a peek when lookup the connection, I was worry it will affect performance but by running benchmark it reported msg rate not changed. 
Yes it just narrows a race, agreed that "if the peer disconnects immediately after peek succeeds" it still will cause failure, that is abnormal use case that the failure should be expected. This patch  can resolve a DAOS normal use case that server keep running but peer finishes and restarts executing with same address.

Probably can with better solution for example re-establish the connection if the disconnect is detected (maybe better to reuse the original AV slot, retry the failed operation automatically so it can be handled inside socket provider and transparent to user, if it is possible).

Reply:
another simple similar use case is nodes A and B, A send RPC request to B, and B will send reply to A. But A only insert B's address to AV at first time and reuse it later. After several RPC round trip, B restart executing. Then A cannot send RPC request to B any more because it already use the cached fi_addr (by daos) and cache connection information (by libfabric).
By this patch, daos still use the cache fi_addr, but in libfabric it will peek to know the connection is broken and re-establish the connection to B, so the RPC communication can continue.

Reply:
After I looked deeper into the code, I don't expect the peek to make a significant difference in performance.  This doesn't occur on each data transfer.

What would help here is to expand the commit message to capture the details listed above.  In this case a short comment in the code would help as well.  For example, "Verify that the existing connection is still usable, and that the peer didn't restart."

Reply:
I have refreshed the patch to expand the commit msg and add the comment as you suggested. thx